lib/puppet/indirector/rest.rb

```diff
@@ -249,7 +250,21 @@ def save(request)
   # to request.do_request from here, thus if we change what we pass or how we
   # get it, we only need to change it here.
   def do_request(request)
-    request.do_request(self.class.srv_service, self.class.server, self.class.port) { |req| yield(req) }
+    response = request.do_request(self.class.srv_service, self.class.server, self.class.port) { |req| yield(req) }
+
+    handle_response(request, response) if response
+
+    response
+  end
+
+  def handle_response(request, response)
+    server_version = response[Puppet::Network::HTTP::HEADER_PUPPET_VERSION]
+    if server_version &&
+       SemanticPuppet::Version.parse(server_version).major < 5 &&
+       Puppet[:preferred_serialization_format] != 'pson'
+      Puppet.warning("Downgrading to PSON for future requests")
+      Puppet[:preferred_serialization_format] = 'pson'
```

Comment:
Actually, it looks like this specific case is broken anyway - a newer agent talking to an older master can't send a report in msgpack:

```
2017-06-12 18:00:14,122 ERROR [qtp1277458017-59] [puppetserver] Puppet Server Error: Could not intern from msgpack: can't convert String into Integer
org/jruby/RubyFixnum.java:1143:in `[]'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction/report.rb:234:in `initialize_from_hash'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction/report.rb:137:in `from_data_hash'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/network/formats.rb:10:in `intern'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/network/format_support.rb:12:in `convert_from'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/network/http/api/indirected_routes.rb:195:in `read_body_into_model'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/network/http/api/indirected_routes.rb:174:in `do_save'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/network/http/api/indirected_routes.rb:48:in `call'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/context.rb:65:in `override'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet.rb:294:in `override'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/network/http/api/indirected_routes.rb:47:in `call'
```